pkg/orchestrator/selection/ranking/available_capacity.go

```diff
@@ -0,0 +1,151 @@
+package ranking
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/bacalhau-project/bacalhau/pkg/lib/math"
+	"github.com/bacalhau-project/bacalhau/pkg/models"
+	"github.com/bacalhau-project/bacalhau/pkg/orchestrator"
+	"github.com/rs/zerolog/log"
+)
+
+// Constants to normalize resource values to a comparable scale.
+const (
+	// Scaling factors for different resource types to normalize them to a comparable scale.
+	memoryScale = 1e6 // Scale memory to megabytes
+	diskScale   = 1e9 // Scale disk to gigabytes
+	gpuScale    = 1   // No scaling needed for GPU as it's already in units
+	cpuScale    = 1   // No scaling needed for CPU as it's already in cores
+
+	// Weights for different resource types. These values are chosen to reflect
+	// the relative importance of each resource type in the context of node capacity.
+	// CPU is considered the most critical resource, followed by memory, disk, and GPU.
+	defaultCPUWeight    = 0.4 // CPU has the highest weight as it directly affects computation speed.
+	defaultMemoryWeight = 0.3 // Memory is crucial for handling large datasets and multitasking.
+	defaultDiskWeight   = 0.2 // Disk space is important but less critical than CPU and memory for most tasks.
+	defaultGPUWeight    = 0.1 // GPU is given the lowest weight as not all tasks require GPU processing.
+
+	// Rank range constants
+	maxAvailableCapacityRank = 80
+	maxQueueCapacityRank     = 20
+)
+
+// AvailableCapacityNodeRanker ranks nodes based on their available capacity and queue used capacity.
+type AvailableCapacityNodeRanker struct{}
+
+// NewAvailableCapacityNodeRanker creates a new instance of AvailableCapacityNodeRanker.
+func NewAvailableCapacityNodeRanker() *AvailableCapacityNodeRanker {
+	return &AvailableCapacityNodeRanker{}
+}
+
+// dynamicWeights calculates the weights for resources based on the job requirements.
+func dynamicWeights(jobRequirements *models.Resources) (float64, float64, float64, float64) {
```

Comment:
probably worth using named returns here, or encapsulate the result in a struct.

Reply:
done in https://github.com/bacalhau-project/bacalhau/pull/4002/commits/c5fbcb9a37ab61cd9bb19f9e02591fec71ebde1d